compact.go

```diff
@@ -0,0 +1,109 @@
+package bbolt
+
+func Compact(dst, src *DB, txMaxSize int64) error {
+	// commit regularly, or we'll run out of memory for large datasets if using one transaction.
+	var size int64
+	tx, err := dst.Begin(true)
+	if err != nil {
+		return err
+	}
+	defer tx.Rollback()
+
+	if err := walk(src, func(keys [][]byte, k, v []byte, seq uint64) error {
+		// On each key/value, check if we have exceeded tx size.
+		sz := int64(len(k) + len(v))
+		if size+sz > txMaxSize && txMaxSize != 0 {
+			// Commit previous transaction.
+			if err := tx.Commit(); err != nil {
+				return err
+			}
+
+			// Start new transaction.
+			tx, err = dst.Begin(true)
+			if err != nil {
+				return err
+			}
+			size = 0
+		}
+		size += sz
+
+		// Create bucket on the root transaction if this is the first level.
+		nk := len(keys)
+		if nk == 0 {
+			bkt, err := tx.CreateBucket(k)
+			if err != nil {
+				return err
+			}
+			if err := bkt.SetSequence(seq); err != nil {
+				return err
+			}
+			return nil
+		}
+
+		// Create buckets on subsequent levels, if necessary.
+		b := tx.Bucket(keys[0])
+		if nk > 1 {
+			for _, k := range keys[1:] {
+				b = b.Bucket(k)
+			}
+		}
+
+		// Fill the entire page for best compaction.
+		b.FillPercent = 1.0
```

Comment:
I wonder whether this line means that the bucket stays forever with the fill factor of 1.0 ? 

E.g. user has default fill factor of pages: 50%... calls compaction, and lands on 100% fill level not only on compacted data (that makes some sense), but also on all future writes.

Reply:
Optional TODO: To verify/fix that.

Reply:
for sequential writes, it is optimal to keep the fill factor to 100%. in the compact/copy case, it should be all sequential writes if i remember correctly.

Reply:
Yes. During compaction it make sense to pack the data in the dense way. 
What I'm afraid of is that this settings stays for this bucket forever (post compation). 
IMHO it should go back to what the user had configured pre-compaction.

Reply:
I miss-assumed the setting is preserved between transactions. But it's scoped to a transaction.